Fix JAX sampling funcs overwriting existing var's dims and coords

jhrcook opened the pull request:
**What is this PR about?**

The original solution to fix Issue #5932 overwrites existing dimensions and coordinate data on the variables when creating the ArviZ InferenceData object. (See the issue for a demonstration of this behavior.) This PR address that by using the `dims` and `coords` in the `idata_kwargs` argument to *update* the extracted dimensions and coordinates.

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Bugfixes / New features
- Fix JAX sampling funcs overwriting existing var's dims and coords.

jhrcook committed:
chore: tidy up existing test for idata kwargs

jhrcook committed:
test: add to test to cause failure

jhrcook committed:
feat: update coords and dims extracted from model with idata_kwargs in JAX samplers

jhrcook committed:
doc: update docstring of JAX samplers with behavior for coords and dims

jhrcook committed:
fix: remove pytest mark used in dev

jhrcook committed:
style: fix typehints for older Python vers

bherwerth commented on pymc/tests/test_sampling_jax.py:
```diff
@@ -199,19 +209,29 @@ def test_idata_kwargs(model_test_idata_kwargs, sampler, idata_kwargs, postproces
+    assert list(x_coords_expected) == list(posterior["x"].coords[x_dim_expected].values)
+
+    assert posterior["z"].dims[2] == "z_coord"
+    assert set(posterior["z"].coords["z_coord"].values) == {"apple", "banana", "orange"}
```
Why do you compare sets here rather than the stronger requirement of equality of lists?

bherwerth commented on pymc/sampling_jax.py:
```diff
@@ -172,6 +172,17 @@ def _get_batched_jittered_initial_points(
+        coords.update(idata_kwargs.pop("coords"))
+    if "dims" in idata_kwargs:
+        dims.update(idata_kwargs.pop("dims"))
+    return None
```
Consider dropping the this line as it's not necessary. Or maybe this is a recommended style, which I am not yet aware of, for a function that should not return anything?

jhrcook replied:
I think it's purely a style choice that I generally do, but will of course follow the style of the repo.

bherwerth commented on pymc/sampling_jax.py:
```diff
@@ -376,6 +388,8 @@ def sample_blackjax_nuts(
     }
 
     posterior = mcmc_samples
+    # Update 'coords' and 'dims' extracted from the model with user 'idata_kwargs'.
```
Consider adding to the comment something like "and drop keys 'coords' and 'dims' from 'idata_kwargs' if present'

bherwerth commented on pymc/sampling_jax.py:
```diff
@@ -596,6 +611,8 @@ def sample_numpyro_nuts(
     }
 
     posterior = mcmc_samples
+    # Update 'coords' and 'dims' extracted from the model with user 'idata_kwargs'.
```
Consider extending the comment as above.

bherwerth reviewed (commented):
I've added some minor comments & questions. Otherwise, looks all good to me.

jhrcook committed:
Merge branch 'pymc-devs:main' into jax-samplers-keeping-dims

jhrcook committed:
style: minor styling changes per code review'

- add details to comments about changing idata_kwargs
- change test to comparing lists instead of sets
- remove an explicit 'return None'

twiecki commented:
Thanks @jhrcook and @bherwerth!
